File: leetcode/groups/0001-0100/[46]全排列.py
# 给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。 
# 
#  
# 
#  示例 1： 
# 
#  
# 输入：nums = [1,2,3]
# 输出：[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
#  
# 
#  示例 2： 
# 
#  
# 输入：nums = [0,1]
# 输出：[[0,1],[1,0]]
#  
# 
#  示例 3： 
# 
#  
# 输入：nums = [1]
# 输出：[[1]]
#  
# 
#  
# 
#  提示： 
# 
#  
#  1 <= nums.length <= 6 
#  -10 <= nums[i] <= 10 
#  nums 中的所有整数 互不相同 
#  
#  Related Topics 数组 回溯 
#  👍 1487 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        length = len(nums)
        if length == 1:
            return [[nums[0]]]      # 不能是[nums], 中途会不断修改同一个nums导致出错
        container = []
        for i in range(length):     # 交换
            nums[i], nums[length - 1] = nums[length - 1], nums[i]
            temp = nums.pop()       # 栈移除末端元素, 剩余元素继续递归, 这种末端元素可以是原有nums中任意元素, 要交换位置
            for subcontainer in self.permute(nums):
                subcontainer.append(temp)
                container.append(subcontainer)
            nums.append(temp)
            nums[i], nums[length - 1] = nums[length - 1], nums[i]
        return container

# leetcode submit region end(Prohibit modification and deletion)


class Solution:
    def permute(self, nums):
        # 不同组合可以视为原始数组nums基础上交换元素实现, 不断增大起始交换点进行递归
        # 参与交换的一对点中, 前面的为start, 后续交换中start增大, 避免重复
        def backtrack(start):
            if start == n:
                res.append(nums[:])
            for i in range(start, n):
                nums[start], nums[i] = nums[i], nums[start]    # 交换
                backtrack(start + 1)
                nums[start], nums[i] = nums[i], nums[start]    # 恢复

        n = len(nums)
        container = []
        backtrack(0)
        return container


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        container = []

        def backtrack(target, temp):    # 没有对输入参数做出修改
            if not target:
                container.append(temp)
            else:
                for i in range(len(target)):
                    backtrack(target[: i] + target[i + 1:], temp + [target[i]])    # 不能用temp.append, 会改变其他地方已有的temp

        backtrack(nums, [])
        return container

File: leetcode/0001-0100/[46]全排列.java
//给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。 
//
// 
//
// 示例 1： 
//
// 
//输入：nums = [1,2,3]
//输出：[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
// 
//
// 示例 2： 
//
// 
//输入：nums = [0,1]
//输出：[[0,1],[1,0]]
// 
//
// 示例 3： 
//
// 
//输入：nums = [1]
//输出：[[1]]
// 
//
// 
//
// 提示： 
//
// 
// 1 <= nums.length <= 6 
// -10 <= nums[i] <= 10 
// nums 中的所有整数 互不相同 
// 
// Related Topics 数组 回溯 
// 👍 1487 👎 0


import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    List<Integer> target;

    public List<List<Integer>> permute(int[] nums) {    // 慢一些
        target = new ArrayList<>(nums.length);
        for (int num: nums) {
            target.add(num);
        }
        return permute(target);
    }

    private List<List<Integer>> permute(List<Integer> target) {
        List<List<Integer>> container = new ArrayList<>();
        if (target.size() == 1) {
            container.add(new ArrayList<>(target));
            return container;
        }
        for (int i = 0; i < target.size(); i++) {
            Collections.swap(target, i, target.size() - 1);
            int temp = target.remove(target.size() - 1);
            for (List<Integer> subcontainer: permute(target)) {
                List<Integer> tempsub = new ArrayList<>(subcontainer);
                tempsub.add(temp);
                container.add(tempsub);
            }
            target.add(temp);
            Collections.swap(target, i, target.size() - 1);
        }
        return container;
    }
}

//leetcode submit region end(Prohibit modification and deletion)


class Solution {
    List<List<Integer>> container = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        List<Integer> target = new ArrayList<>(nums.length);
        for (int num : nums) {
            target.add(num);
        }
        backtrack(target, new ArrayList<>());
        return container;
    }

    private void backtrack(List<Integer> target, List<Integer> cache) {
        if (target.isEmpty()) {
            container.add(cache);    // 不能直接用cache, 后面不断修改cache会导致已经加入的cache改变
        }
        else {
            for (int i = 0; i < target.size(); i++) {
                List<Integer> subcache = new ArrayList<>(cache.subList(0, cache.size()));
                List<Integer> subtarget = new ArrayList<>(target.subList(0, target.size()));
                subcache.add(target.get(i));
                subtarget.remove(i);
                backtrack(subtarget, subcache);
            }
        }
    }
}


class Solution {
    List<List<Integer>> container = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        List<Integer> aim = new ArrayList<>(nums.length);
        for (int num: nums) {
            aim.add(num);
        }
        backtrack(0, aim);
        return container;
    }

    private void backtrack(int start, List<Integer> target) {
        if (start == target.size()) {
            container.add(new ArrayList<>(target));
        }
        for (int i = start; i < target.size(); i++) {
            Collections.swap(target, start, i);
            backtrack(start + 1, target);
            Collections.swap(target, start, i);
        }
    }
}
